activation/nipost.go

```diff
@@ -377,64 +356,134 @@ func (nb *NIPostBuilder) submitPoetChallenge(
 	}
-	g.Wait()
-	close(errChan)
 
-	allInvalid := true
-	for err := range errChan {
-		if err == nil {
-			allInvalid = false
-			continue
+	registrationsMap := make(map[string]nipost.PoETRegistration)
+	for _, reg := range registrations {
+		registrationsMap[reg.Address] = reg
+	}
+
+	existingRegistrationsMap := make(map[string]nipost.PoETRegistration)
+	missingRegistrationsPoETAddresses := make([]string, 0)
+
+	for addr := range nb.poetProvers {
+		if val, ok := registrationsMap[addr]; ok {
+			existingRegistrationsMap[addr] = val
+		} else {
+			missingRegistrationsPoETAddresses = append(missingRegistrationsPoETAddresses, addr)
 		}
+	}
+
+	existingRegistrations := maps.Values(existingRegistrationsMap)
+	if len(missingRegistrationsPoETAddresses) == 0 {
+		return existingRegistrations, nil
+	}
 
-		nb.logger.Warn("failed to submit challenge to poet", zap.Error(err), log.ZShortStringer("smesherID", nodeID))
-		if !errors.Is(err, ErrInvalidRequest) {
-			allInvalid = false
+	now := time.Now()
+
+	if curPoetRoundStartDeadline.Before(now) {
+		if len(existingRegistrations) == 0 {
+			if len(registrations) == 0 {
+				// no existing registration at all, drop current registration challenge
+				err = ErrATXChallengeExpired
+			} else {
+				// no existing registration for given poets set
+				err = ErrNoRegistrationForGivenPoetFound
+				nb.logger.Warn("revert poet configuration to previous is needed immediately",
+					zap.Error(err), log.ZShortStringer("smesherID", nodeID))
+			}
+			return nil, fmt.Errorf(
+				"%w: poet round has already started at %s (now: %s)",
+				err,
+				curPoetRoundStartDeadline,
+				now,
+			)
 		}
+		return existingRegistrations, nil
 	}
-	if allInvalid {
-		nb.logger.Warn("all poet submits were too late. ATX challenge expires", log.ZShortStringer("smesherID", nodeID))
-		return ErrATXChallengeExpired
+
+	// send registrations to missing addresses
+	signature := signer.Sign(signing.POET, challenge)
+	prefix := bytes.Join([][]byte{signer.Prefix(), {byte(signing.POET)}}, nil)
+
+	submitCtx, cancel := context.WithDeadline(ctx, curPoetRoundStartDeadline)
+	defer cancel()
+
+	g, ctx := errgroup.WithContext(submitCtx)
+	submittedRegistrationsChan := make(chan nipost.PoETRegistration, len(missingRegistrationsPoETAddresses))
+
+	for _, addr := range missingRegistrationsPoETAddresses {
+		if client, ok := nb.poetProvers[addr]; ok {
+			g.Go(func() error {
+				registration, err := nb.submitPoetChallenge(
+					ctx, nodeID,
+					poetProofDeadline,
+					client, prefix, challenge, signature)
+				if err != nil {
+					nb.logger.Warn("failed to submit challenge to poet",
+						zap.Error(err),
+						log.ZShortStringer("smesherID", nodeID))
+				} else {
+					submittedRegistrationsChan <- registration
+				}
+				return nil
+			})
+		} else {
+			nb.logger.Warn("poet service not found",
+				log.ZShortStringer("smesherID", nodeID),
+				zap.String("poetAddr", addr))
+		}
+	}
```

Comment:
Then this can be simplified:
```suggestion
	for _, client := range missingRegistrations {
		g.Go(func() error {
			registration, err := nb.submitPoetChallenge(
				ctx, nodeID,
				poetProofDeadline,
				client, prefix, challenge, signature,
			)
			if err != nil {
				nb.logger.Warn("failed to submit challenge to poet",
					zap.Error(err),
					log.ZShortStringer("smesherID", nodeID),
				)
			} else {
				submittedRegistrationsChan <- registration
			}
			return nil
		})
	}
```